slack_ros/scripts/slack_service.py

```diff
@@ -0,0 +1,170 @@
+
+    def register_callback(self, fkt):
+        if callable(fkt):
+            self.callbacks.append(fkt)
+        else:
+            sys.stderr("register_callback: arg is not callable")
+
+    def unregister_callback(self, fkt):
+        self.callbacks.remove(fkt)
+
+    def send_slack(self, data):
+        self.client.api_call(
+            "chat.postMessage", channel=self.channel, text=data,
+            username=self.username, icon_emoji=':robot_face:'
+        )
+
+    def send_slack_file(self, data):
+        with open(data.data, 'rb') as file:
+            requests.post('https://slack.com/api/files.upload', 
+                files={'file': ['File '+data.data, file]}, params={
+                'token': self.token,
+                'channels': self.channel
+            })
+
+    def send_zmq(self, data):
+        self.socket_pub.send_string(data)
+
+class SystemInterpreter(object):
+    def __init__(self):
+        self.system = System()
+        self.bringup = Bringup()
+
+    def slack_cb(self, msg):
+        cmd = "self."+msg[2:]
+        try:
+            return eval(cmd)
+        except Exception as e:
+            print "SystemInterpreter: error eval '%s': %s"%(cmd, e.message)
+            raise e
+
+class InterpreterMediator(object):
+    def __init__(self, slack_srv):
+        self.sys_int = SystemInterpreter()
+        self.slack_srv = slack_srv
+        self.slack_srv.register_callback(self.slack_callback)
+
+    def __del__(self):
+        self.slack_srv.register_callback(self.slack_callback)
+
+    def slack_callback(self, data):
+        #first check if it is a command
+        msg = unescape(data)
+
+        #it is a interpreter command
+        if msg.startswith('> '):
+            #first pass it to system interpreter
+            try:
+                self.slack_srv.send_slack(self.sys_int.slack_cb(msg))
+            except Exception as e:
+                #if system interpreter issues an error
+                #check if ros is running and send it to ros interpreter
+                #since we do not know if ros interpreter was successfull at 
+                #this stage, we just forward the ros interpreter output
+                if is_roscore_up():
+                    self.slack_srv.send_zmq(data)
+                #if ros is not running
+                else:
+                    self.slack_srv.send_slack("Error exec command. Roscore not running and system reported error: %s"%e.message)
+        #no interpreter command just send it to zmq socket
+        else:
+            self.slack_srv.send_zmq(data)
+
+if __name__ == '__main__':
+    if len(sys.argv) != 4:
+        sys.stderr.write("Wrong num command line arguments!")
+        sys.exit(os.EX_CONFIG)
+
+    signal.signal(signal.SIGINT, signal_handler)
+
+    slack_srv = SlackService(sys.argv[1], sys.argv[2], sys.argv[3])
+    mediator = InterpreterMediator(slack_srv)
+    
+    while not SIGINT:
+
+        try:
+            rospy.get_master().getPid()
```

Comment:
this makes it rosdependend again...

Reply:
sure, to be more precise, the node should be roscore independent. It's ok to use rospy here